Skip grid refresh after running a navigational script

crypto-rsa opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #5834.

## Proposed changes

- methods executing a command now return a new `struct CommandStatus` instead of a simple bool. Besides specifying whether the command was actually executed, it also contains a flag indicating whether the revision grid should be refreshed.
- navigational scripts no longer cause the grid refresh

## Test methodology <!-- How did you ensure quality? -->

- Manual testing (running commands both from context menu and using a hotkey)


## Test environment(s) <!-- Remove any that don't apply -->

- git version 2.20.1.windows.1
- Windows 10

<!-- Mention language, UI scaling, or anything else that might be relevant -->


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../../blob/master/contributors.txt).

RussKie commented:
@crypto-rsa please send a separate PR with your sign off to contributors.txt

crypto-rsa commented:
> However it looks like the only place we may require a refresh is in `return new CommandStatus(true, !scriptInfo.RunInBackground);`... I.e. no scripts will refresh the grid.
> Am I reading it correctly?

Not quite. See `new CommandStatus(true, plugin.Execute(eventArgs))` at line 110.

RussKie commented on GitUI/Script/ScriptRunner.cs:
```diff
@@ -115,6 +116,11 @@ private static bool RunScript(IWin32Window owner, GitModule module, ScriptInfo s
+                if (revisionGrid == null)
+                {
+                    return false;
+                }
```
Should this check be significantly higher?
`revisionGrid` is used in other places in the method in a way that would result in a NRE... E.g. line:109

crypto-rsa replied:
Good point. I was able to produce a NRE for a `plugin:` script. However, in the future there might be other types of scripts which would not require the `revisionGrid` instance, so I'd rather not abort running the script if it is `null` in the general case. Instead, I'd just do the null check for the `plugin:` scripts the same way I did for the `navigateTo:` scripts. Is that OK?

crypto-rsa replied:
Hmm, after giving it a second thought, should we really not allow a script to run without the revision grid? It might not actually need it. We should just make sure it doesn't crash when creating the `GitUIEventArgs` instance (at line 109).

RussKie replied:
> should we really not allow a script to run without the revision grid?

No, we should not. Scripts must be able to run without the revision grid - #6135 is a perfect example of the use case.

crypto-rsa replied:
For the `plugin:` scripts, the revision grid is only required to get a `GitUICommands` instance (through its `UICommands` property). Is it possible to get it in a different way? Or should we just pass `null` instead into the `GitUIEventArgs` constructor?

RussKie replied:
`NavigateTo` scripts must only be available in the revision grid context, since it is a command performed on the grid itself.
We should be able to supply the `UICommands` instance to all scripts since it has a number of useful events and properties. 
I will need to investigate when I have time. Feel free to offer your vision.

RussKie commented:
The build is failing because if tests

vbjay commented:
The goal is to expand scripts to other areas.

On Thu, Feb 14, 2019, 11:26 AM Pavel Kotrč <notifications@github.com> wrote:

> *@crypto-rsa* commented on this pull request.
> ------------------------------
>
> In GitUI/Script/ScriptRunner.cs
> <https://github.com/gitextensions/gitextensions/pull/6219#discussion_r256909016>
> :
>
> > @@ -115,6 +116,11 @@ private static bool RunScript(IWin32Window owner, GitModule module, ScriptInfo s
>
>              if (command.StartsWith(NavigateToPrefix))
>              {
> +                if (revisionGrid == null)
> +                {
> +                    return false;
> +                }
>
> Hmm, after giving it a second thought, should we really not allow a script
> to run without the revision grid? It might not actually need it. We should
> just make sure it doesn't crash when creating the GitUIEventArgs instance
> (at line 109).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6219#discussion_r256909016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsQtawW5RXEi6xRR67G6IIInzznTSks5vNY5IgaJpZM4alm6D>
> .
>

crypto-rsa commented:
Also I just had an idea we could show a message box when an attempt is made to run a navigational script without the revision grid (for example through a hotkey when the diff control is focused). Right now, the script is just silently aborted and the user has no idea what happened. However, I don't know what is the GE policy on using message boxes. They are sometimes considered harmful to the workflow.

RussKie commented:
I think from a user's perspective that would constitute to a poor UX.

Imagine you run a script that instead of a result shows a popup saying you
need to run it having focused the revision grid, because the developers did
a poor job procuring the necessary context.

On Fri, Feb 15, 2019, 7:47 PM Pavel Kotrč <notifications@github.com wrote:

> Also I just had an idea we could show a message box when an attempt is
> made to run a navigational script without the revision grid (for example
> through a hotkey when the diff control is focused). Right now, the script
> is just silently aborted and the user has no idea what happened. However, I
> don't know what is the GE policy on using message boxes. They are sometimes
> considered harmful to the workflow.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6219#issuecomment-463956129>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXuuOEOoxxpdliT-mxzRPIH89_QtEks5vNnQ2gaJpZM4alm6D>
> .
>

RussKie commented:
Here's something we can do to pass `GitUICommands`:
```diff
diff --git a/GitUI/Script/ScriptRunner.cs b/GitUI/Script/ScriptRunner.cs
index e5315d92a..752529009 100644
--- a/GitUI/Script/ScriptRunner.cs
+++ b/GitUI/Script/ScriptRunner.cs
@@ -12,7 +12,7 @@ public static class ScriptRunner
     {
         /// <summary>Tries to run scripts identified by a <paramref name="command"/>
         /// and returns true if any executed.</summary>
-        public static bool ExecuteScriptCommand(IWin32Window owner, GitModule module, int command, RevisionGridControl revisionGrid = null)
+        public static bool ExecuteScriptCommand(IWin32Window owner, GitModule module, int command, IGitUICommands uiCommands, RevisionGridControl revisionGrid = null)
         {
             var anyScriptExecuted = false;
 
@@ -20,7 +20,7 @@ public static bool ExecuteScriptCommand(IWin32Window owner, GitModule module, in
             {
                 if (script.HotkeyCommandIdentifier == command)
                 {
-                    RunScript(owner, module, script.Name, revisionGrid);
+                    RunScript(owner, module, script.Name, uiCommands, revisionGrid);
                     anyScriptExecuted = true;
                 }
             }
@@ -28,7 +28,7 @@ public static bool ExecuteScriptCommand(IWin32Window owner, GitModule module, in
             return anyScriptExecuted;
         }
 
-        public static bool RunScript(IWin32Window owner, GitModule module, string scriptKey, RevisionGridControl revisionGrid)
+        public static bool RunScript(IWin32Window owner, GitModule module, string scriptKey, IGitUICommands uiCommands, RevisionGridControl revisionGrid)
         {
             if (string.IsNullOrEmpty(scriptKey))
             {
```

crypto-rsa commented:
> Here's something we can do to pass `GitUICommands`:

Thanks @RussKie! I've used the `IGitUICommands` instance you've passed in for the `plugin:` scripts which now work even without a valid reference to `RevisionGridControl`.
I also fixed the failing test and rebased on top of `master`.

RussKie commented:
Thank you
